rpc/gnmi/gnmi-history.md

```diff
@@ -0,0 +1,87 @@
+# gNMI History Extension
+
+**Contributors:** Justin Costa-Roberts, Aaron Beitch, Mike Fink
+
+**Date:** August 4, 2020
+
+**Version:** 0.1.0
+
+# 1. Purpose
+
+In certain applications it is necessary not only to retrieve current data from a
+data tree, but also to allow querying historical data. The gNMI History
+extension enables clients both to retrieve data at a specific time in the past,
+and to request all updates applied to a data tree between two specified times.
+
+# 2. Definition
+
+A `History` message is embedded in an Extension message in the
+`SubscribeRequest` proto.
+
+## 2.1 Proto
+
+```
+message TimeRange {
+  int64 start = 1;
+  int64 end = 2;
+}
+
+message History {
+  oneof request {
+    int64 time = 1;
+    TimeRange range = 2;
+  }
+}
+```
+
+## 2.2 SubscribeRequest handling
+
+The History extension may be applied only to `ONCE` and `STREAM` subscriptions.
+Requests for historical data can be either of two types: a snapshot at a
+specified time, or a set of updates between two times. The `History` message's
+request field is a `oneof` field whose type indicates which of these request
+types applies.
+
+In all cases, times are represented as int64s interpreted as nanoseconds since
+the Unix Epoch.
+
+### 2.2.1 Snapshot requests
+
+If the `time` field is set, the request is for a snapshot. The subscription mode
+for a time request must be `ONCE`; otherwise the server must return an
+`INVALID_ARGUMENT` error. The server must transmit the data indicated in the
+`SubscribeRequest` as it existed on the server at the specified time. If the
+specified time indicates a future time at the point the server receives the
+request, the server may either return an `UNIMPLEMENTED` error or choose to wait
+until the specified time has been reached.
+
+### 2.2.2 Stream requests
+
+If the `range` field is set, the request is for the set of updates that occurred
+during the half-open interval between the specified start and end times. Such a
+request may return multiple updates per leaf, so the subscription mode for a
+range request must be `STREAM`. If the mode of a range request is not `STREAM`,
+the server should return an `INVALID_ARGUMENT` error. For the paths specified in
+the `SubscriptionList`, the server must transmit all updates it's aware of that
+occurred at the start time or later and before the end time specified in the
+`TimeRange`, and then return. Whether the server should transmit the specified
+data tree as it existed at the start time is determined as usual by the
+`updates_only` field on the `SubscriptionList` message.
+
+If `TimeRange`'s start time is later than its end time, the server must return
```

Comment:
I feel like this going to be a little more complicated in the expectations from the service providing the stream...

If you have something that both going to serve historical data as well as current data how the data is muxed into those streams is likely to be complicated.

If you want it easy on the client - you would say start at the beginning and just play our the stream from history and then once done start sending all future updates from the live stream but that could have huge buffering problems on the server side of things.

The other option is to allow the streams to be mixed and just play the updates out of order but that would like cause clients problems as they wouldn't know when the "history" stream is actually done or they would just drop all non latest updates as currently for the clients i use - receiving out of order updates -- the old ones are just dropped since they have been replaced by new data anyway

Reply:
My reading of the gNMI spec is that updates needn't be buffered until the sync_response message has been sent:

"When the target has transmitted the initial updates for all paths specified within the subscription, a SubscribeResponse message with the sync_response field set to true MUST be transmitted to the client to indicate that the initial transmission of updates has concluded. This provides an indication to the client that all of the existing data for the subscription has been sent at least once. For STREAM subscriptions, such messages are not required for subsequent updates."

So that should cover this case as well.

Reply:
I think that in the scenario  that you have a client that wants to take both "historical and future state", then the behaviour that seems logical (and the lowest implementation overhead) is:

 * play all updates starting at `start_time`
 * once this has completed, send current version of the tree

For a history-consuming client, there is already a need to parse more than a single update to a path within the stream. For example, if we are drawing some graph, then we insert a point at all received datapoints. The buffering would ensure no discontinuities in this graph - but I suspect such a client should rather implement:

- history query for $start_time -> $now
- live subscription

if it is interested in both live and historical data. Or simply a history query if it already knows the time bounds.

Reply:
I'm not sure I understand the first case you're describing--there's no combination of start and end time that should produce updates beginning at a historical start time and then the whole current tree. If updates_only isn't set on a TimeRange subscription, the server should send the tree as it existed at the specified start time, and then incremental updates from then up until the specified end.

In the second case, I agree that clients may just do a historical query + a normal subscription, but are you saying that something about the spec should change, or just pointing out that clients always have the option of implementing open-ended historical queries this way?

Reply:
I'm not sure that the "history + whole current tree" is forbidden in the current gNMI specification, right? I agree that the optimal approach is not to re-send the current tree (since this might be sending duplicates), but there's no specific requirement that a server MUST NOT ever send the same update twice (in-fact, this is required for a solely SAMPLE subscription). I was just pointing out that Marcus' concern on implementation complexity might be mitigated in this manner.

For the latter, I was also intending to support this implementation -- and was just trying to highlight that I think client implementation doesn't necessarily need to change here (for a use case where having not seen a particular set of updates that were dropped/suppressed by the server whilst replaying the history would have a material impact).

Reply:
How would you specify "history + whole current tree"? The only way to get the full tree with a `STREAM` subscription is to set `updates_only` to `false`. If the History extension is used, this will return the tree at the `TimeRange`'s specified `start` time, and then any updates that occurred between that time and the specified `end` time. Thus the only way to get the current tree with the History extension is to specify a `start` time of time.Now, which means no historical data. So history + whole current tree is not possible.

My understanding of Marcus' concern is that it's about whether, in servicing a `STREAM` subscription with `start` < `now` < `end`, the server should buffer updates that occur between `now` and `end` until the initial state has been synchronized (i.e., until the server has sent a `sync_response`), or just transmit them as soon as they occur. That is, I think Marcus' point is about update ordering, not duplication. And by my reading, the spec doesn't prohibit sending current updates before the initial tree has synchronized.

Reply:
Apologies, I think we're talking past each other here. I am not saying that we need a way to be able to have the updates plus the whole current tree, I'm saying that if an implementation honoured a history approach by implementing this way then this would not be in violation of the spec.

Regarding the buffering, I agree with you - there is no requirement for ordering in the stream in the current gNMI. I don't see that we should introduce this now.